ADBDEV-4789: Fix an incorrect memory assumption for hash join statistics

bandetto opened the pull request:
Fix an incorrect memory assumption for hash join statistics

When a hashjoin's hashtable runs out of allowed memory, it divides the hash
table into batches, and stores each batch to disk. When using `EXPLAIN ANALYZE`,
an additional description is created for the number of batches and the memory
they used. (With normal `EXPLAIN`, the request is not executed, and accordingly,
batches are not created). To store information about the disk space occupied by
a batch, the `HashJoinTableStats` structure is used, which has a `batchstats`
member of type `HashJoinBatchStats`, which is an array of structures for each
batch.

Usually, the number of batches is calculated at the beginning of hashjoin, but
sometimes the calculation algorithm may miss. Then, the
`ExecHashIncreaseNumBatches()` is used, which increases the current number of
batches, multiplying the previous one by two, and allocating additional memory
to the structures that need it.

Each structure of type `HashJoinBatchStats` takes 80 bytes, opposed to 8 bytes
for a pointer. Since memory for `batchstats` member is allocated for the
structures themselves, and not pointers to them, the size of `realloc()` call
for `batchstats` in `ExecHashIncreaseNumBatches()` can become more than
`MaxAllocSize` and cause an error.

In the beginning of `ExecHashIncreaseNumBatches()`, there is a check that
assumes that each item in any of the reallocated arrays would not be more than 8
bytes (`sizeof(void *)`). `oldnbatch` is equal to `nbatch / 2`. When we take the
size of the `HashJoinBatchStats` into account, we may notice that this check is
not correct specifically for this allocation, and is off by 10 times of the
value of the faulty `repalloc()` call.

The issue is that `batchstats` grows faster than other arrays containing
pointers. So this check does not prevent the faulty statistics batches
reallocation. When we are nearing `MaxAllocSize` with
`nbatch * sizeof(stats->batchstats[0])`, and try to reallocate the `batchstats`
member, we will encounter an error, because allocation size requested will be
more than `MaxAllocSize`.

Use `repalloc_huge()` instead of `repalloc()` for statistics, to allow
requesting memory up to:
`MaxAllocSize / ((sizeof(void *) * 2)` = `67 108 863`;
`67 108 863 * 80` = `5 368 709 040`, which is slightly less than 5GB.

On a 32-bit machines, `nbatch * sizeof(stats->batchstats[0])` could exceed
`MaxAllocHugeSize`, which is `2 ^ 32` = 4GB. Add a separate check and stop if
`nbatch > MaxAllocHugeSize / sizeof(HashJoinBatchStats)` to avoid integer
overflow.

On 64-bit machines, maximum allowed amount of batches is unchanged.

bandetto committed:
Revert "Allocate memory for explain batches separately"

This reverts commit 4b6273b8bd2011c5b3d974f4966781c084c26d78.

bandetto committed:
Use repalloc_huge instead of repalloc

bandetto committed:
Allocate memory for explain batches separately

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/62045

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/977661

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/977662

RekGRpth commented on src/backend/executor/nodeHash.c:
```diff
@@ -768,11 +768,20 @@ ExecHashIncreaseNumBatches(HashJoinTable hashtable)
 	if (stats && stats->nbatchstats < nbatch)
 	{
-		Size		sz = nbatch * sizeof(stats->batchstats[0]);
+		Size		sz = nbatch * sizeof(HashJoinBatchStats);
```
Is `sizeof(HashJoinBatchStats)` equal to `sizeof(stats->batchstats[0])` or not?

bandetto replied:
Yes, it is. `stats->batchstats[0]` is of type `HashJoinBatchStats`

RekGRpth replied:
then remove unnecessary changes

bandetto replied:
Done. I just thought this was slightly confusing, because everything else in this function uses types for `sizeof`.

bandetto committed:
HashJoinBatchStats -> stats->batchstats[0]

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/62101

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/980215

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/980216

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/980209

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/980210

RekGRpth commented on src/backend/executor/nodeHash.c:
```diff
@@ -770,8 +770,17 @@ ExecHashIncreaseNumBatches(HashJoinTable hashtable)
+		 */
 		stats->batchstats =
-			(HashJoinBatchStats *) repalloc(stats->batchstats, sz);
+			(HashJoinBatchStats *) repalloc_huge(stats->batchstats, sz);
```
How about additional [check for 32-bit machine](https://github.com/arenadata/gpdb/commit/263865a48973767ce8ed7b7788059a38a24a9f37#diff-5b735c65568e935c0bdf768116c93c1266ac7b4c516a06fb1aa847b4e093cff1R1049-R1051) (where MaxAllocHugeSize ~ 4Gb)?

bandetto replied:
Done

bandetto committed:
Add an overflow check for 32 bit systems

On a 32-bit machines, `nbatch * sizeof(stats->batchstats[0])` could exceed
MaxAllocHugeSize, which is SIZE_MAX / 2. Stop if `nbatch > MaxAllocHugeSize /
sizeof(HashJoinBatchStats)` to avoid integer overflow.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/62201

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/984620

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/984614

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/984621

RekGRpth commented:
provide last commit message

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/62245

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/986017

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/986018

RekGRpth commented:
> provide last commit message

I meant the commit message that will go into the resulting patch.

bimboterminator1 commented on src/backend/executor/nodeHash.c:
```diff
@@ -731,6 +731,10 @@ ExecHashIncreaseNumBatches(HashJoinTable hashtable)
 		return;
 
+	/* avoid repalloc_huge overflow on 32 bit systems */
+	if (stats && oldnbatch > MaxAllocHugeSize / (sizeof(HashJoinBatchStats) * 2))
+		return;
```
1. Description remarks

>So this check is off by 10 times of the actual value of the
faulty repalloc() call

The first check, which does not prevent the faulty statistics batches reallocation, is also related for the code below and looks quite guarding for that case. In other words, you should change the description, to make the reader understand, that this check is not erroneous, the issue is that `batchstats` grows faster than the array of pointers.
https://github.com/arenadata/gpdb/blob/665598ae127d0c52affddb0c73ea7a47adaa3671/src/backend/executor/nodeHash.c#L761-L765

Finally, patches are currently merged by andr-sokolov, and the final commit message is taken from PR description. So, please, rewrite it with proper 80 alignment. The description should fully cover the situation. Starting from introduction to the issue, ending with the solution proposal and its validity proof.

2.  10 GB estimate is too high, the maximum number of batches is `2^26 = 67 108 864` ( because `MaxAllocSize / (sizeof(void *) * 2)` equals `67 108 863`). It means that maximum memory is ~6GB ? Probably should also write this in the comments, and specifiy that this estimate is taken for MaxAllocSize = 1G

bandetto replied:
1. Done.
2. Since `MaxAllocSize / ((sizeof(void *) * 2)` = `67 108 863`, `67 108 863 * 80` = `5 368 709 040`, which is slightly less than 5GB, so you are right.

bandetto committed:
Fix the estimation and clarify that it's for MaxAllocSize = 1GB

bandetto committed:
Rewrap the comment to 80 columns

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/62415

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/992700

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-4789

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/62523

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/997908

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/997909

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/997906
